storage: change allocator metrics to counters

tamird opened the pull request:
Changes allocator metrics to record actions taken by the allocator,
rather than attempting to record future actions by the allocator, which
is inherently racy.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/11733)
<!-- Reviewable:end -->

mberhault commented on pkg/storage/metrics.go:
```diff
@@ -61,7 +61,7 @@ var (
 
-	// Replication metrics.
+	// Allocator metrics.
 	metaReplicaAllocatorNoopCount       = metric.Metadata{Name: "ranges.allocator.noop"}
```
since you're poking around, could you add Help strings to these metrics?

petermattis commented:
Using gauges for these metrics allowed us to see when replicate queue had a large backlog of work. For example, when a node is down for 5min it is declared dead causing a spike in `removedead` work for the replicate queue. How will we see that spike after this change?

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-:-KXuncbF7pB6qNYLJEP6:budyonj)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Also, what problems did using gauges cause? Just wondering what was the initial motivation for this PR.

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-:-KXur2d7OhjhjIMAEawb:b-wh1cxj)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reworked this as discussed offline. @petermattis PTAL. I'll add the new
replicate queue counters in a subsequent PR.

On Thu, Dec 1, 2016 at 9:49 AM, Peter Mattis <notifications@github.com>
wrote:

> Also, what problems did using gauges cause? Just wondering what was the
> initial motivation for this PR.
> ------------------------------
>
> Review status: 0 of 6 files reviewed at latest revision, 1 unresolved
> discussion, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-:-KXur2d7OhjhjIMAEawb:b-wh1cxj>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/11733#issuecomment-264192014>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPO3nNmXuRoVLrLoe2SDu5t3eB1rOks5rDt6QgaJpZM4LA7rz>
> .
>

petermattis commented:
Looks better.

---

Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replicate_queue.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv8SKZg9MUW6XAm4kv:-KXv8SK_crvgdEsoqwEA:b-sggsb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L206)):*
> ```Go
> 	// quorum.
> 	deadReplicas := rq.allocator.storePool.deadReplicas(desc.RangeID, desc.Replicas)
> 	{
> ```

This block isn't helping readability. Perhaps this should be a `hasQuorum(desc, len(deadReplicas)) bool` function.

---

*[pkg/storage/replicate_queue.go, line 267 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv8wAKLH0oXnKXpfKL:-KXv8wAKLH0oXnKXpfKM:blcha5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L267)):*
> ```Go
> 			// out of situations where this store is overfull and yet holds all the
> 			// leases.
> 			transferred, err := rq.transferLease(ctx, repl, desc, zone, false)
> ```

Comment what `false` means here.

---

*[pkg/storage/replicate_queue.go, line 302 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv9-3aOs757MUkdgXO:-KXv9-3aOs757MUkdgXP:b-fryqwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L302)):*
> ```Go
> 			// We require the lease in order to process replicas, so
> 			// repl.store.StoreID() corresponds to the lease-holder's store ID.
> 			transferred, err := rq.transferLease(ctx, repl, desc, zone, true)
> ```

Comment what `true` means here.

---

*[pkg/storage/store.go, line 3740 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv9ax3ygXSLopsdDSv:-KXv9ax3ygXSLopsdDSw:bnj1utl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/store.go#L3740)):*
> ```Go
> 			switch action, _ := s.allocator.ComputeAction(zoneConfig, desc); action {
> 			case AllocatorAdd, AllocatorRemoveDead:
> 				underreplicatedRangeCount++
> ```

I realize this is what we discussed, but now that I look at the code I'm wondering if we should use the liveness info instead of `AllocatorRemoveDead`. Specifically, don't even bother with `ComputeAction` and instead compute the number of under replicated ranges below using the liveness table.

---

*[pkg/storage/store.go, line 3955 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXvA44HfHsuELwnkEX-:-KXvA44HfHsuELwnkEX0:bbza5qh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/store.go#L3955)):*
> ```Go
> 			log.Error(rep.AnnotateCtx(context.TODO()), err)
> 		} else {
> 			switch action, _ := s.allocator.ComputeAction(zoneConfig, desc); action {
> ```

`ComputeAction` might be doing too much here. Should this be looking at the desired number of replicas in the zone config vs `len(desc.Replicas)`? Should this be consulting the liveness info?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-:-KXvAL-ZpgMG5JuF5r9O:b-3fm243)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/storage/replicate_queue.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv8SKZg9MUW6XAm4kv:-KXvAxti0ZJjSoKd5pxV:b-yuhzpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L206)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This block isn't helping readability. Perhaps this should be a `hasQuorum(desc, len(deadReplicas)) bool` function.
</blockquote></details>

I added the block to reduce the scope of these variables. I see now that I randomly reversed the two lines, which I've now reverted.

---

*[pkg/storage/replicate_queue.go, line 267 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv8wAKLH0oXnKXpfKL:-KXvAl58ASfSUMzgpEMX:bdbdyzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L267)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Comment what `false` means here.
</blockquote></details>

I figured it was OK to omit because it's in the same file, but done.

---

*[pkg/storage/replicate_queue.go, line 302 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv9-3aOs757MUkdgXO:-KXvAnoNb8TTmqbpbtCS:bdbdyzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L302)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Comment what `true` means here.
</blockquote></details>

I figured it was OK to omit because it's in the same file, but done.

---

*[pkg/storage/store.go, line 3740 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv9ax3ygXSLopsdDSv:-KXvDr7L69PMn0nRVQQe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/store.go#L3740)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I realize this is what we discussed, but now that I look at the code I'm wondering if we should use the liveness info instead of `AllocatorRemoveDead`. Specifically, don't even bother with `ComputeAction` and instead compute the number of under replicated ranges below using the liveness table.
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 3955 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXvA44HfHsuELwnkEX-:-KXvDrdEBvR3ymxYNjHN:b-6jde89) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/store.go#L3955)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`ComputeAction` might be doing too much here. Should this be looking at the desired number of replicas in the zone config vs `len(desc.Replicas)`? Should this be consulting the liveness info?
</blockquote></details>

Reverted this to use the metric (now the new one).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/metrics.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXsBo_8Y-AeB0yv_Rrc-r1-65:-KXvF7-fShAcPJ-mR8zj:b-h0rkv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ba330f2aa4869d43440c752fd2267f13b7d0e8/pkg/storage/metrics.go#L65)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

since you're poking around, could you add Help strings to these metrics?
</blockquote></details>

Don't forget to address this.

---

*[pkg/storage/replicate_queue.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXv8SKZg9MUW6XAm4kv:-KXvEXPPh20qRXLavuC3:b2ly6hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f89d692ac19ae386baa944f02b5e16c441e4d434/pkg/storage/replicate_queue.go#L206)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I added the block to reduce the scope of these variables. I see now that I randomly reversed the two lines, which I've now reverted.
</blockquote></details>

Right, I get that it reduces the scope of those variables, but that wasn't hurting anything. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-:-KXvF5bAkWsrfd0N85EJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/metrics.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXsBo_8Y-AeB0yv_Rrc-r1-65:-KXvK-lAWnsb05KXLtOG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ba330f2aa4869d43440c752fd2267f13b7d0e8/pkg/storage/metrics.go#L65)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Don't forget to address this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/metrics.go, line 66 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXvLL6cT3PgK3z7IJX7:-KXvLL6d7JxzY15F46Pf:b-ilo7zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7242de06aaa137a0c05fbe7239f5e63905c7676/pkg/storage/metrics.go#L66)):*
> ```Go
> 		Help: "Number of ranges with a quorum of live replicas"}
> 	metaUnderReplicatedRangeCount = metric.Metadata{Name: "ranges.underreplicated",
> 		Help: "Number of ranges with fewer than the configured number of live replicas"}
> ```

That reads a bit weird. How about: `Number of ranges with fewer live replicas than the target replication`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11733)*
<!-- Sent from Reviewable.io -->

tamird commented:
Agreed, will change.

On Thu, Dec 1, 2016 at 1:29 PM, Peter Mattis <notifications@github.com>
wrote:

> Review status: 0 of 10 files reviewed at latest revision, 3 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/storage/metrics.go, line 66 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11733#-KXvLL6cT3PgK3z7IJX7:-KXvLL6d7JxzY15F46Pf:b-ilo7zr>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/b7242de06aaa137a0c05fbe7239f5e63905c7676/pkg/storage/metrics.go#L66>):*
>
> 		Help: "Number of ranges with a quorum of live replicas"}
> 	metaUnderReplicatedRangeCount = metric.Metadata{Name: "ranges.underreplicated",
> 		Help: "Number of ranges with fewer than the configured number of live replicas"}
>
> That reads a bit weird. How about: Number of ranges with fewer live
> replicas than the target replication?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11733>*
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/11733#issuecomment-264253091>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPJSNRmCwildCAUPFHZ7EWKSe_PAXks5rDxHxgaJpZM4LA7rz>
> .
>
